### ISSUE TITLE
Manage strings for gravity-force-lab

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/inverse-square-law-common#26]
This repo has code  for gravity-force-lab, gravity-force-lab-basics, and coulombs-law. We would also like to consolidate shared strings. Since gravity-force-lab has already been published we were concerned that translations would break if we move shared strings to this repo.

@jbphet is this correct? Do you have a recommendation for how to handle this?

**Comment by jessegreenberg** [comment]
This issue came from https://github.com/phetsims/inverse-square-law-common/issues/26

**Comment by jbphet** [comment]
Indeed, this is a situation for which we don't have a very clean solution.  The general idea would be to move all the strings from GFL to common code, and do the same thing with all translations, and do a maintenance release of GFL.  While this is happening, we'd probably want to take down rosetta, or at least remove GFL from the list of translatable sims for the duration of the effort.

Also, we'd probably want to create a script to move the strings from the sim to the common repo.

**Comment by jessegreenberg** [comment]
Thanks @jbphet, it sounds like the long term solution will be to do those steps.  When we deploy a sim using this repo, we can coordinate to go through those steps. Removing assignment, but this must be done before coulombs-law or gravity-force-lab-basics is deployed.

**Comment by mbarlow12** [comment]
@jessegreenberg Is the implication here that we'll need to move all strings from GFL and GFLB into ISLC regardless of whether they're specific to the sim or not?

**Comment by jessegreenberg** [comment]
@mbarlow12 I don't think so, it is more that translated versions of GFL will lose their translations or maybe break loudly because strings that were in gravity-force-lab-strings_xx.json are now in inverse-square-law-strings_xx.json. So we need to make sure that the movement is done for all locales.

**Comment by zepumph** [comment]
>Removing assignment, but this must be done before coulombs-law or gravity-force-lab-basics is deployed.

CL is already published. @jessegreenberg can you please give the state of this issue. What needs to be done here before GFLB can be published?

**Comment by jessegreenberg** [comment]
This problem came up during developer meeting on May 30. From the developer meeting doc:
> ✓JO: Do we have a good way of moving strings (that may be duplicated in published sims) into common code? It seems like quite a tricky problem.
CM: Our policy in the past (and present?) has been not to move them, and require translators to re-translate.
JB: Yes, having translators re-translate is the current process.  We have scripts to move strings but the scripts aren’t ready for prime-time.
AP: Make an issue, put it in rosetta.  Mark it deferred.  Can be looked at for rosetta 2.0

Here is the issue that was created: https://github.com/phetsims/rosetta/issues/212.

So my understanding is that the duplication of strings in gravity-force-lab and inverse-square-law-common is expected at this time. Strings have been moved out of gravity-force-lab-strings_en.json and into inverse-square-law-common-strings_en.json, and I fear that may cause problems. @jbphet can you please comment? Do we need to make sure that gravity-force-lab-strings_en.json has all the strings that existed at the time of its last publication?

**Comment by jbphet** [comment]
@jessegreenberg asked:

> Strings have been moved out of gravity-force-lab-strings_en.json and into inverse-square-law-common-strings_en.json, and I fear that may cause problems. @jbphet can you please comment? Do we need to make sure that gravity-force-lab-strings_en.json has all the strings that existed at the time of its last publication?

No, we don't need to make sure that the the English string file on master has all the strings from the last publication.  I can explain the details if needed, but the consequences of this move will be that the existing translations of GFLB will lose the translations of the strings that were moved the next time it is published from master.  This seems like a bit of a disservice to our translators, and it is possible to copy the translated strings, it's just a bit of a manual process.  I could to it in ~2 hrs start to finish.  @ariel-phet  - is it worth it?

**Comment by ariel-phet** [comment]
@jbphet I think it is worth it - is that something you could perhaps document when you do it? Or have someone like Saurabh or such ride along and document so if it needs to be done in the future you don't have to be responsible? Or if documentation already exists, feel free to pass along the task to an intern.

**Comment by zepumph** [comment]
Does this need to be done before publication? I believe so, but wanted to be sure. For now note that this is one of the last two items to do before going to RC for GFLB (see also https://github.com/phetsims/gravity-force-lab-basics/issues/166). Marking as high priority.

**Comment by jbphet** [comment]
@jessegreenberg - I started taking a look at this, but I'm a bit confused.  In https://github.com/phetsims/inverse-square-law-common/commit/62c9373a32a8a463cf1ba114e60e75c5acf3c648, it looks like the strings were added to the inverse-square-law-repo, but the strings that were added in that commit don't appear to have been removed from `gravity-force-lab-strings_en.json`.  Am I missing something here?

**Comment by terracoda** [comment]
@jbphet, the "mass 1" and "mass 2" strings are the names for the mass spheres in GFL:Basics. In GFL regular, the mass spheres' names are "m1" and "m2".

Does that help?

**Comment by terracoda** [comment]
Oh, @jbphet, I see the labels are in this file for GFLBasics
https://github.com/phetsims/gravity-force-lab-basics/blob/master/gravity-force-lab-basics-strings_en.json

So back to your original question in https://github.com/phetsims/inverse-square-law-common/issues/26#issuecomment-508851313

**Comment by jessegreenberg** [comment]
Sorry for the confusion, it looks like the strings in gravity-force-lab-strings_en.json could be cleaned up a bit. It would help to know exactly what changed in the strings file.

Here is the diff between 1.2 release branch and master for gravity-force-lab-strings_en.json:
<details>

```diff
--- C:\Users\Jesse\Documents\Development\phetsims\gravity-force-lab\temp.json Mon Jul  8 12:15:25 2019
+++ C:\Users\Jesse\Documents\Development\phetsims\gravity-force-lab\gravity-force-lab-strings_en.json Wed Jun 19 15:29:34 2019
@@ -7,9 +7,6 @@
   },
   "mass2": {
     "value": "Mass 2"
-  },
-  "showValues": {
-    "value": "Show Values"
   },
   "units.kg": {
     "value": "kg"
@@ -23,16 +20,43 @@
   "mass2Abbreviated": {
     "value": "m2"
   },
-  "pattern_0value_1units": {
-    "value": "{0} {1}"
+  "moveSpheresHeading": {
+    "value": "Move Spheres"
   },
-  "force-description-pattern-target_source": {
-    "value": "Force on {0} by {1}"
+  "moveSphereLabel": {
+    "value": "Move sphere"
   },
-  "force-description-pattern-target_source_value": {
-    "value": "Force on {0} by {1} = {2} N"
+  "moveInSmallerSteps": {
+    "value": "Move in smaller steps"
   },
-  "constantRadius": {
-    "value": "Constant Radius"
+  "moveInLargerSteps": {
+    "value": "Move in larger steps"
+  },
+  "jumpToLeft": {
+    "value": "Jump to left"
+  },
+  "jumpToRight": {
+    "value": "Jump to right"
+  },
+  "changeMassHeading": {
+    "value": "Change Mass"
+  },
+  "increaseMass": {
+    "value": "Increase mass"
+  },
+  "decreaseMass": {
+    "value": "Decrease mass"
+  },
+  "changeMassInLargerSteps": {
+    "value": "Change mass in larger steps"
+  },
+  "changeMassInSmallerSteps": {
+    "value": "Change mass in smaller steps"
+  },
+  "jumpToMaximumMass": {
+    "value": "Jump to maximum mass"
+  },
+  "jumpToMinimumMass": {
+    "value": "Jump to minimum mass"
   }
 }
```
</details>

So the following string keys were removed from gravity-force-lab-strings_en.json and put in inverse-square-law-common-strings_en.json, and possibly renamed in the process:
showValues -> forceValues
pattern_0value_1units -> pattern_0value_1units
force-description-pattern-target_source -> forceOnObjectByOtherObjectPattern
force-description-pattern-target_source_value -> forceOnObjectByOtherObjectWithUnitsPattern
constantRadius -> constantSize
 
String patterns that moved were changed to use double curly brace notation.

Before continuing with the process @jbphet proposed the other common strings should be removed from gravity-force-lab-strings_en.json, which are `mass1`, `mass2`, and `units.kg`.

**Comment by jbphet** [comment]
I've reviewed the information above and @jessegreenberg just had a zoom conversation about this.  It turns out that not only have the strings been moved, they have also in most cases been renamed (as noted in the comment just above), and in some cases the pattern type has been changed (from the older single-brace style to the newer double-brace style).  This makes the process of moving them much more complex, and the time to deal with it would far exceed the two hours that I estimated above.  At this point, I recommend that we do **not** try to move the previously translated strings, and that we instead notify translators the next time we publish "Gravity Force Lab" off of master that their translation will need to be updated.  I'll log an issue for this in the repo.

Assigning to @ariel-phet in case he has any objections or alternative thoughts.

**Comment by jessegreenberg** [comment]
> the other common strings should be removed from gravity-force-lab-strings_en.json, which are mass1, mass2, and units.kg.

@jbphet also confirmed that these can be removed from gravity-force-lab since they were moved to inverse-square-law-common and that was done in the above commit.

**Comment by zepumph** [comment]
I discussed the above changes over with @jessegreenberg and it seems like the Mass1/2  strings should be in GFL instead.

**Comment by zepumph** [comment]
>Assigning to @ariel-phet in case he has any objections or alternative thoughts.

@ariel-phet if you feel like it is acceptable to leave the babel data alone, and require new translations upon next release, I think this issue is ready to close.

**Comment by ariel-phet** [comment]
All sounds good to me. Closing
